gnovm/pkg/gnolang/preprocess.go

```diff
@@ -3779,16 +3791,486 @@ func assertTypeDeclNoCycle2(store Store, last BlockNode, x Expr, stack *[]Name,
 // type expressions, which must get preprocessed for inner
 // composite type eliding to work.
 func findUndefined(store Store, last BlockNode, x Expr) (un Name) {
-	return findUndefined2(store, last, x, nil)
+	return findUndefined2(store, last, x, nil, true)
+}
+
+// finds the next undefined identifier and returns it if it is global
+func findUndefined2SkipLocals(store Store, last BlockNode, x Expr, t Type) Name {
+	name := findUndefinedGlobal(store, last, x, t)
```

Comment:
as the base logic of the PR, please add comments to this function. to better illustrate the `searching logic`: 
for my understanding: searching from initialized global -> uninitialized global decls -> local...